lib/usd/pxrUsdPreviewSurface/usdPreviewSurfacePlugin.cpp

```diff
@@ -26,78 +30,92 @@
 #include <maya/MDrawRegistry.h>
 
 #include <pxr/base/tf/envSetting.h>
+#include <pxr/base/tf/stringUtils.h>
+#include <pxr/usdImaging/usdImaging/tokens.h>
 
-PXR_NAMESPACE_USING_DIRECTIVE
+PXR_NAMESPACE_OPEN_SCOPE
 
 namespace {
-const MString _RegistrantId("mayaUsd");
-int _registrationCount = 0;
-
-// Name of the plugin registering the preview surface class.
-MString _registrantPluginName;
-
+TfToken::Set _registeredTypeNames;
 }
-
-PXR_NAMESPACE_OPEN_SCOPE
-
 /* static */
-MStatus
-PxrMayaUsdPreviewSurfacePlugin::initialize(MFnPlugin& plugin)
+MStatus PxrMayaUsdPreviewSurfacePlugin::initialize(
+    MFnPlugin&     plugin,
+    const MString& typeName,
+    MTypeId        typeId,
+    const MString& registrantId)
 {
-    // If we're already registered, do nothing.
-    if (_registrationCount++ > 0) {
-        return MS::kSuccess;
+    TfToken typeNameToken(typeName.asChar());
+    if (_registeredTypeNames.count(typeNameToken) > 0) {
+        TF_CODING_ERROR("Trying to register typeName %s more than once", typeNameToken.GetText());
+        return MStatus::kFailure;
     }
+    _registeredTypeNames.insert(typeNameToken);
 
-    _registrantPluginName = plugin.name();
+    MString drawDbClassification(
+        TfStringPrintf("drawdb/shader/surface/%s", typeName.asChar()).c_str());
+    MString fullClassification(
+        TfStringPrintf("shader/surface:shader/displacement:%s", drawDbClassification.asChar())
+            .c_str());
 
     MStatus status = plugin.registerNode(
-        PxrMayaUsdPreviewSurface::typeName,
-        PxrMayaUsdPreviewSurface::typeId,
+        typeName,
+        typeId,
         PxrMayaUsdPreviewSurface::creator,
         PxrMayaUsdPreviewSurface::initialize,
         MPxNode::kDependNode,
-        &PxrMayaUsdPreviewSurface::fullClassification);
+        &fullClassification);
     CHECK_MSTATUS(status);
 
-    status =
-        MHWRender::MDrawRegistry::registerSurfaceShadingNodeOverrideCreator(
-            PxrMayaUsdPreviewSurface::drawDbClassification,
-            _RegistrantId,
-            PxrMayaUsdPreviewSurfaceShadingNodeOverride::creator);
+    status = MHWRender::MDrawRegistry::registerSurfaceShadingNodeOverrideCreator(
+        drawDbClassification, registrantId, PxrMayaUsdPreviewSurfaceShadingNodeOverride::creator);
     CHECK_MSTATUS(status);
 
+    UsdMayaShaderWriterRegistry::Register(
+        typeNameToken,
```

Comment:
Registering a node will automatically register a reader/writer for file import/export.